File: IV - TRS on equity basket/Python code/TRS on equity basket.py
# -*- coding: utf-8 -*-
"""
Created on Sat Mar  4 13:55:04 2023

@author: moureaux pierre
"""

import numpy as np
from matplotlib import pyplot as plt
import scipy
import scipy.linalg

"Market datas"
M = 3
S0 = np.arange(100,100+M,1)
sigma0 = np.full(M,0.1)
r0 = 0.01

"TRS datas"
T = 1
TRSpaymentSchedule = (25,50,75,99)
sTRS = 0.0006

"Model parameters"
kappa = np.full(M,0.1)
theta = np.full(M,0.1)
lamb = np.full(M,0.1)
nu = 0.1
etha = 0.1
alpha = 0.02
deltaT = 0.01
lT = int(T/deltaT)
nbSimul = 1000

"Model correlations"
corrSS = np.matrix([[1, 0.2,0.1], [0.2, 1,0.3],[0.1, 0.3,1]])
L = scipy.linalg.cholesky(corrSS, lower=True)
rhoSsigma = np.full(M,0.1)
rhoSr = np.full(M,0.1)

def pathsGeneration(nbSimul,lT,S0,sigma0,r0,kappa,theta,lamb,nu,etha,alpha,rhoSsigma,rhoSr):
    paths = np.zeros(nbSimul,dtype=dict)
    for i in range(nbSimul):
        x = np.random.normal(0, 1, (M,lT))
        phiS = np.dot(L,x)
        phitild=np.random.normal(0, 1, (M,lT))
        phihat=np.random.normal(0, 1, lT)
        S = np.zeros((M,lT))
        sigma = np.zeros((M,lT))
        r = np.zeros(lT)
        intr = np.zeros(lT)
        S[:,0]=S0
        sigma[:,0]=sigma0
        r[0]=r0
        intr[0]=r[0]
        for k in range(lT-1):
            for j in range(M):
                r[k+1] = (nu-etha*r[k])*deltaT + np.sqrt(alpha*r[k])*(rhoSr[j]*np.sqrt(deltaT)*phiS[j,k]+np.sqrt(1-rhoSr[j]*rhoSr[j])*np.sqrt(deltaT)*phihat[k])+r[k]
                intr[k+1]=intr[k]+r[k+1]
                sigma[j,k+1] = kappa[j]*(theta[j]-sigma[j,k])*deltaT + lamb[j]*np.sqrt(sigma[j,k])*(rhoSsigma[j]*np.sqrt(deltaT)*phiS[j,k]+np.sqrt(1-rhoSsigma[j]*rhoSsigma[j])*np.sqrt(deltaT)*phitild[j,k])+sigma[j,k]
                S[j,k+1] = S[j,k]*(r[k]*deltaT+np.sqrt(sigma[j,k]*deltaT)*phiS[j,k]) + S[j,k]
        paths[i] = {"Asset": S, "Volatility": sigma, "Interest rate": r, "Cum. interest rate": intr}
    return paths

def alphaBetaTRS(TRSpaydate,prevTRSpaydate, paths, nbSimul):
    alphak = 0
    betak = 0
    for i in range(nbSimul):
        sum1 = 0
        sum2 = 0
        for j in range(M):
            sum1 += paths[i]['Asset'][j,TRSpaydate]
            sum2 += paths[i]['Asset'][j,prevTRSpaydate]
        res1 = (sum1-sum2)/(np.exp(paths[i]['Cum. interest rate'][TRSpaydate]))
        res2 = (sum2)/(np.exp(paths[i]['Cum. interest rate'][TRSpaydate]))
        alphak += res1
        betak += res2
    alphak /= nbSimul
    betak /= nbSimul
    return [alphak,betak]

paths = pathsGeneration(nbSimul,lT,S0,sigma0,r0,kappa,theta,lamb,nu,etha,alpha,rhoSsigma,rhoSr)

def TRS(TRSpaymentSchedule, sTRS, paths,nbSimul):
    res = 0
    alpha = 0
    beta = 0
    for s in TRSpaymentSchedule:
        alpha += alphaBetaTRS(s,TRSpaymentSchedule[TRSpaymentSchedule.index(s)-1],paths,nbSimul)[0]
        beta += alphaBetaTRS(s,TRSpaymentSchedule[TRSpaymentSchedule.index(s)-1],paths,nbSimul)[1]*(s-TRSpaymentSchedule[TRSpaymentSchedule.index(s)-1])
    res = alpha - sTRS*beta
    impliedRate = alpha/beta
    return [res, impliedRate]

print("TRS market value is",TRS(TRSpaymentSchedule, sTRS, paths,nbSimul)[0])
print("TRS implied rate is",TRS(TRSpaymentSchedule, sTRS, paths,nbSimul)[1])

File: 4 - TRS on equity basket/C++ code/TRS on equity basket.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <unordered_map>
#include <random>
#include <math.h>
#include <any>
#include <Eigen/Cholesky>
#include <EigenRand/EigenRand>
#include <numeric>

typedef std::vector<std::unordered_map<std::string, std::any>> typePaths;

//Market datas
auto M = 3;

//TRS datas
auto T = 1.0;
std::vector<double> TRSpaymentSchedule = { 0,25, 50, 75, 99 };
auto sTRS = 0.01;

//Parameters
auto deltaT = 0.01;
auto lT = static_cast<int>(T / deltaT);
auto nbSimul = 100;

typePaths pathsGeneration(int nbSimul, int lT, double r0, const std::vector<double>& kappa,
    const std::vector<double>& theta, const std::vector<double>& lamb, double nu, double etha,
    double alpha, const Eigen::MatrixXd& L, const std::vector<double>& S0, const std::vector<double>& sigma0,
    std::vector<double>& rhoSsigma,const std::vector<double>& rhoSr)
{
    typePaths paths;
    for (auto i = 0; i != nbSimul; i++)
    {
        std::default_random_engine generator;
        Eigen::MatrixXd x = Eigen::Rand::normal<Eigen::MatrixXd>(M, lT, generator, 0.0, 1.0);
        Eigen::MatrixXd phiS = L.cast<double>() * x.cast<double>();
        Eigen::MatrixXd phitild = Eigen::Rand::normal<Eigen::MatrixXd>(M, lT, generator, 0.0, 1.0);
        std::vector<double> phihat;
        for (auto k = 0; k != lT; k++)
        {
            phihat.emplace_back(std::normal_distribution<double>{0, 1}(generator));
        }
        Eigen::MatrixXf S(M,lT);
        Eigen::MatrixXf sigma(M,lT);
        std::vector<double> r(lT);
        std::vector<double> intr(lT);
        for (auto k = 0; k != M;k++)
        {
            S(k, 0) = S0[k];
            sigma(k, 0) = sigma0[k];
        }
        r[0] = r0;
        intr[0] = r[0];
        for (auto k = 0; k != (lT - 1); k++)
        {
            for (auto j = 0; j != M; j++)
            {
                r[k + 1] = (nu - etha * r[k]) * deltaT + sqrt(alpha * r[k]) * (rhoSr[j] * sqrt(deltaT) * phiS(j,k) + 
                    sqrt(1 - rhoSr[j] * rhoSr[j]) * sqrt(deltaT) * phihat[k]) + r[k];
                intr[k + 1] = intr[k] + r[k + 1];
                sigma(j,k+1) = kappa[j] * (theta[j] - sigma(j,k)) * deltaT + lamb[j] * sqrt(sigma(j,k)) * (rhoSsigma[j] * sqrt(deltaT) * phiS(j,k) +
                    sqrt(1 - rhoSsigma[j] * rhoSsigma[j]) * sqrt(deltaT) * phitild(j,k)) + sigma(j,k);
                S(j,k+1) = S(j,k) * (r[k] * deltaT + sqrt(sigma(j,k) * deltaT) * phiS(j,k)) + S(j,k);
            }
        }
        std::unordered_map<std::string, std::any> interRes = { {"Asset",S},{"Volatility",sigma},{"Interest rate",r},{"Cum. interest rate",intr} };
        paths.emplace_back(interRes);
    }
    return paths;
}

std::vector<double> alphaBetaTRS(int TRSpaydate, int prevTRSpaydate, const typePaths& paths, int nbSimul)
{
    auto alphak = 0.0;
    auto betak = 0.0;
    for (auto i = 0; i != nbSimul; i++)
    {
        auto sum1 = 0.0;
        auto sum2 = 0.0;
        for (auto j = 0; j != M; j++)
        {
            sum1 += std::any_cast<Eigen::MatrixXf>(paths[i].at("Asset"))(j,TRSpaydate);
            sum2 += std::any_cast<Eigen::MatrixXf>(paths[i].at("Asset"))(j, prevTRSpaydate);
        }
        auto res1 = (sum1 - sum2) / (exp(std::any_cast<std::vector<double>>(paths[i].at("Cum. interest rate"))[TRSpaydate]));
        auto res2 = (sum2) / (exp(std::any_cast<std::vector<double>>(paths[i].at("Cum. interest rate"))[TRSpaydate]));
        alphak += res1;
        betak += res2;
    }
    alphak /= nbSimul;
    betak /= nbSimul;
    std::vector<double> res;
    res.emplace_back(alphak);
    res.emplace_back(betak);
    return res;
}

int getIndex(const std::vector<double>& v, int K)
{
    auto it = find(v.begin(), v.end(), K);
    if (it != v.end())
    {
        int index = it - v.begin();
        return index;
    }
    else
    {
        return 1;
    }
}

std::vector<double> TRS(const std::vector<double>& TRSpaymentSchedule, double sTRS, const typePaths& paths, int nbSimul)
{
    auto alpha = 0.0;
    auto beta = 0.0;
    for (auto s = TRSpaymentSchedule.begin(); s != TRSpaymentSchedule.end() - 1; s++)
    {
        auto payDate = getIndex(TRSpaymentSchedule, *s) + 1;
        auto prevPayDate = getIndex(TRSpaymentSchedule, *s);
        auto abRes = alphaBetaTRS(TRSpaymentSchedule[payDate], TRSpaymentSchedule[prevPayDate], paths, nbSimul);
        alpha += abRes[0];
        beta += abRes[1] * (TRSpaymentSchedule[payDate] - TRSpaymentSchedule[prevPayDate]);
    }
    auto resTemp = alpha - sTRS * beta;
    auto impliedRate = alpha / beta;
    std::vector<double> res;
    res.emplace_back(resTemp);
    res.emplace_back(impliedRate);
    return res;
}

int main()
{
    //Market datas
    std::vector<double> S0(M);
    std::iota(S0.begin(), S0.end(), 100);
    std::vector<double> sigma0(M);
    std::fill(sigma0.begin(), sigma0.end(), 0.1);
    auto r0 = 0.05;

    //Model datas
    std::vector<double> kappa(M);
    std::vector<double> theta(M);
    std::vector<double> lamb(M);
    std::fill(kappa.begin(), kappa.end(), 0.1);
    std::fill(theta.begin(), theta.end(), 0.1);
    std::fill(lamb.begin(), lamb.end(), 0.1);
    auto nu = 0.1;
    auto etha = 0.1;
    auto alpha = 0.02;

    //Model correlations"
    Eigen::Matrix3f corrSS;
    corrSS << 1, 0.2, 0.1,
              0.2, 1, 0.3,
              0.1, 0.3, 1;
    Eigen::MatrixXd L(corrSS.cast<double>().llt().matrixL());
    std::vector<double> rhoSsigma(M);
    std::vector<double> rhoSr(M);
    std::fill(rhoSsigma.begin(), rhoSsigma.end(), 0.1);
    std::fill(rhoSr.begin(), rhoSr.end(), 0.1);

    //Pricing
    auto paths = pathsGeneration(nbSimul, lT, r0, kappa,
        theta, lamb, nu, etha,
        alpha, L, S0, sigma0,
        rhoSsigma, rhoSr);
    auto TRSres = TRS(TRSpaymentSchedule, sTRS, paths, nbSimul);
    std::cout << "TRS market value is" << TRSres[0] << std::endl;
    std::cout << "TRS implied rate is" << TRSres[1] << std::endl;
}